### ISSUE TITLE
Not able to connect to Che Workspaces from remote computer on Mac

### ISSUE THREAD

**Issue by darrinsimmons** [eclipse-che/che#3092]
I have installed Che several different ways and can connect to Che, start workspaces and access them from my Mac.  I can access Che and start workspaces from my PC.  I am unable to access workspaces from my PC.
1: git clone https://github.com/eclipse/che ; cd che ; ./che.sh start
2: docker run --rm -t -v /var/run/docker.sock:/var/run/docker.sock -e CHE_IP=192.168.128.19 -e CHE_HOST_IP=192.168.128.19 -e CHE_DOCKER_MACHINE_HOST_EXTERNAL=192.168.129.19 eclipse/che:5.0.0-latest start

**Reproduction Steps:**

1. Install Che via either method above.  I did not export any environment variables or set any che.properties.
2. Create loopback alias to 192.168.65.2
3. Access Che from same machine.  Everything is fine.
4. Access Che from different machine.  Can't access workspaces.

**Expected behavior:**
I expected to be able to access workspaces from a remote machine.

**Observed behavior:**
I can access Che and start workspaces from a remote machine.  I am unable to access workspaces from a remote machine.

**Che version:**    5.0.0-latest   
**OS and version:** MacOS Sierra Version 10.12.1  
**Docker version:** 1.12.3(13776) 
**Che install:**   Docker Container   

**Additional information:**

This has never worked.  When I debug on my PC inside Chrome, I see Chrome trying to access localhost instead of 192.168.128.19 when trying to connect to the workspace agents.

This is 100% reproducible.

**Comment by TylerJewell** [comment]
Ok - so I think many of those environment variables may not be necessary.  Can you remove all of them and just verify first that you can access Che and use it on your Mac from "http://localhost:<port>"?  If, yes, then that just means that your MAC is not advertising an externally accessible IP address.

If you think it is, do you have an IP address which is externally pingable from another computer?

**Comment by darrinsimmons** [comment]
Thanks @TylerJewell for the attention.

When I:
docker stop che-server
docker rm che-server
docker run --rm -t -v /var/run/docker.sock:/var/run/docker.sock eclipse/che:5.0.0-latest start
INFO: ECLIPSE CHE: Found image eclipse/che-server:5.0.0-latest
INFO: ECLIPSE CHE: Starting container...
INFO: ECLIPSE CHE: Server logs at "docker logs -f che-server"
INFO: ECLIPSE CHE: Server booting...
INFO: ECLIPSE CHE: Booted and reachable
INFO: ECLIPSE CHE: Ver: 5.0.0-M7
INFO: ECLIPSE CHE: Use: http://192.168.65.2:8080
INFO: ECLIPSE CHE: API: http://192.168.65.2:8080/swagger

I am able to access Che, start and access workspaces from my Mac from:
http://localhost:8080
http://192.168.65.2:8080
http://192.168.128.19:8080

I am able to access Che and start workspaces from my PC from:
http://192.168.128.19:8080 (recall from above that I have the loopback alias for 192.168.65.2)
I am not able to access workspaces from my PC.

I am not able to access Che or workspaces from my PC from:
http://192.168.65.2:8080 as this address is not routable outside of my Mac.

192.168.128.19 is reachable from my PC as can be seen by the fact that my PC is able to view Che on http://192.168.128.19:8080.

Again, in Chrome developer console on my PC, I see network connections failing to the wsagent on localhost when I try and access a workspace.
http://localhost:33252/wsagent/ext/project
So the Che workspace page running in the browser is trying to connect to localhost instead of the IP address of the docker container running the workspace.  My expectation would be to have all communication to wsagents use the $CHE_IP address and let docker ensure packets get routed to the right private IP addresses on the host instead of having the wsagents hardcode localhost.

The workspaces running on my Mac do not have routable IP addresses.  Here are the IP addresses for the workspace:
Darrins-MacBook-Pro:dhs-cloud dsimmons$ docker exec -i -t bca358083034 /bin/bash
user@bca358083034:/projects$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
3: gre0@NONE: <NOARP> mtu 1476 qdisc noop state DOWN group default qlen 1
    link/gre 0.0.0.0 brd 0.0.0.0
4: gretap0@NONE: <BROADCAST,MULTICAST> mtu 1462 qdisc noop state DOWN group default qlen 1000
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
5: ip_vti0@NONE: <NOARP> mtu 1428 qdisc noop state DOWN group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
6: ip6_vti0@NONE: <NOARP> mtu 1500 qdisc noop state DOWN group default qlen 1
    link/tunnel6 :: brd ::
7: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1
    link/sit 0.0.0.0 brd 0.0.0.0
8: ip6tnl0@NONE: <NOARP> mtu 1452 qdisc noop state DOWN group default qlen 1
    link/tunnel6 :: brd ::
9: ip6gre0@NONE: <NOARP> mtu 1448 qdisc noop state DOWN group default qlen 1
    link/gre6 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00 brd 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
289: eth0@if290: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link 
       valid_lft forever preferred_lft forever
291: eth1@if292: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:13:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.19.0.2/16 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe13:2/64 scope link 
       valid_lft forever preferred_lft forever
user@bca358083034:/projects$ 

Thanks again for looking into this.

**Comment by TylerJewell** [comment]
@eivantsov @l0rd - can one of you assist?

**Comment by l0rd** [comment]
@darrinsimmons Please try this:

1/ Remove the loopback alias (Che works without that alias now)

``` sh
export DOCKER_VM_IP=$(docker run --rm --net host alpine sh -c "ip a show eth0" | grep 'inet ' | cut -d/ -f1 | awk '{ print $2}')
sudo ifconfig lo0 $DOCKER_VM_IP delete
```

2/ Make sure you have latest Che images:

``` sh
docker pull eclipse/che-server:5.0.0-latest
docker pull eclipse/che:5.0.0-latest
```

3/ Start Che using the following syntax

``` sh
docker run --rm -t -v /var/run/docker.sock:/var/run/docker.sock \
                               -e CHE_DOCKER_MACHINE_HOST_EXTERNAL=192.168.129.19 \
                               eclipse/che:5.0.0-latest start
```

4/ Do the following checks:
- [ ] You can access Che dashboard from your Mac browser (http://192.168.129.19:8080/)
- [ ] You can create a workspace from your Mac browser
- [ ] If you open Chrome developers tools you see that Che is not using `localhost` anymore but 192.168.129.19 as the workspace address
- [ ] You can access Che dashboard from your PC browser (http://192.168.129.19:8080/)
- [ ] You can create a workspace from your PC

5/ If some of these checks fail please copy/paste the output of 

``` sh
docker run --rm -t -v /var/run/docker.sock:/var/run/docker.sock \
                               -e CHE_DOCKER_MACHINE_HOST_EXTERNAL=192.168.129.19 \
                               eclipse/che:5.0.0-latest 
```

**Comment by darrinsimmons** [comment]
Thanks Mario.  The results are the same as before.  I can do the following:

Mac
192.168.128.19:8080 - Connect to Che, Start Workspace, Connect to Workspace
localhost:8080 - Connect to Che, Start Workspace, Connect to Workspace

PC
192.168.128.19:8080 - Connect to Che, Start Workspace

When connecting to 192.168.128.19 from my Mac (Which is 192.168.128.19), below are all the localhost references from within Chrome-developertools-netwoork.  I have omitted all network connections to 192.168.128.19.

http://localhost:33420/wsagent/ext/project http://localhost:33420/wsagent/ext/project
http://localhost:33420/wsagent/ext/ http://localhost:33420/wsagent/ext/
http://localhost:33420/wsagent/ext/languageserver/supported http://localhost:33420/wsagent/ext/languageserver/supported
ws://localhost:33418/pty ws://localhost:33418/pty
http://localhost:33420/wsagent/ext/project-type http://localhost:33420/wsagent/ext/project-type 
http://localhost:33420/wsagent/ext/project/tree/nodejs-hello-world?depth=2&includeFiles=true http://localhost:33420/wsagent/ext/project/tree/nodejs-hello-world?depth=2&includeFiles=true
ws://localhost:33420/wsagent/ext/ws ws://localhost:33420/wsagent/ext/ws
http://localhost:33420/wsagent/ext/project/tree/hello-world?depth=1&includeFiles=true http://localhost:33420/wsagent/ext/project/tree/hello-world?depth=1&includeFiles=true
http://localhost:33420/wsagent/ext/project/tree/hello-world?depth=1&includeFiles=true http://localhost:33420/wsagent/ext/project/tree/hello-world?depth=1&includeFiles=true
http://localhost:33420/wsagent/ext/project/file/nodejs-hello-world/app/app.js http://localhost:33420/wsagent/ext/project/file/nodejs-hello-world/app/app.js
http://localhost:33420/wsagent/ext/project/file/hello-world/README.md http://localhost:33420/wsagent/ext/project/file/hello-world/README.md
http://localhost:33420/wsagent/ext/java/compiler-settings/all http://localhost:33420/wsagent/ext/java/compiler-settings/all
http://localhost:33420/wsagent/ext/languageserver/registered http://localhost:33420/wsagent/ext/languageserver/registered
http://localhost:33420/wsagent/ext/project/file/hello-world/index.html http://localhost:33420/wsagent/ext/project/file/hello-world/index.html
ws://localhost:33420/wsagent/websocket/2113334070 ws://localhost:33420/wsagent/websocket/2113334070

When connecting to 192.168.128.19 from my PC, here are the localhost references from within Chrom-developertools-network:
http://localhost:33420/wsagent/ext/project http://localhost:33420/wsagent/ext/project
http://localhost:33420/wsagent/ext/ http://localhost:33420/wsagent/ext/
http://localhost:33420/wsagent/ext/languageserver/supported http://localhost:33420/wsagent/ext/languageserver/supported
I assume the remainder are not called because they are based on responses from the 3 above.

Here are the results from starting cha-server and the corresponding logs:
Darrins-MacBook-Pro:dhs-cloud dsimmons$ docker run --rm -t -v /var/run/docker.sock:/var/run/docker.sock -e CHE_DOCKER_MACHINE_HOST_EXTERNAL=192.168.129.19 eclipse/che:5.0.0-latest start
INFO: ECLIPSE CHE: Found image eclipse/che-server:5.0.0-latest
INFO: ECLIPSE CHE: Starting container...
INFO: ECLIPSE CHE: Server logs at "docker logs -f che-server"
INFO: ECLIPSE CHE: Server booting...
INFO: ECLIPSE CHE: Booted and reachable
INFO: ECLIPSE CHE: Ver: 5.0.0-M7
INFO: ECLIPSE CHE: Use: http://192.168.65.2:8080 http://192.168.65.2:8080/
INFO: ECLIPSE CHE: API: http://192.168.65.2:8080/swagger http://192.168.65.2:8080/swagger
Darrins-MacBook-Pro:dhs-cloud dsimmons$ docker logs -f che-server
Using embedded assembly...
Using embedded che.properties... Copying template to /home/user/che/conf.
2016-11-17 16:09:27,270[main]             [INFO ] [o.a.c.s.VersionLoggerListener 89]    - Server version:        Apache Tomcat/8.0.38
2016-11-17 16:09:27,272[main]             [INFO ] [o.a.c.s.VersionLoggerListener 91]    - Server built:          Oct 6 2016 20:51:55 UTC
2016-11-17 16:09:27,273[main]             [INFO ] [o.a.c.s.VersionLoggerListener 93]    - Server number:         8.0.38.0
2016-11-17 16:09:27,273[main]             [INFO ] [o.a.c.s.VersionLoggerListener 95]    - OS Name:               Linux
2016-11-17 16:09:27,273[main]             [INFO ] [o.a.c.s.VersionLoggerListener 97]    - OS Version:            4.4.27-moby
2016-11-17 16:09:27,274[main]             [INFO ] [o.a.c.s.VersionLoggerListener 99]    - Architecture:          amd64
2016-11-17 16:09:27,274[main]             [INFO ] [o.a.c.s.VersionLoggerListener 101]   - Java Home:             /usr/lib/jvm/java-1.8-openjdk/jre
2016-11-17 16:09:27,274[main]             [INFO ] [o.a.c.s.VersionLoggerListener 103]   - JVM Version:           1.8.0_101-b13
2016-11-17 16:09:27,275[main]             [INFO ] [o.a.c.s.VersionLoggerListener 105]   - JVM Vendor:            Oracle Corporation
2016-11-17 16:09:27,275[main]             [INFO ] [o.a.c.s.VersionLoggerListener 107]   - CATALINA_BASE:         /home/user/che/tomcat
2016-11-17 16:09:27,275[main]             [INFO ] [o.a.c.s.VersionLoggerListener 109]   - CATALINA_HOME:         /home/user/che/tomcat
2016-11-17 16:09:27,276[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.util.logging.config.file=/home/user/che/tomcat/conf/logging.properties
2016-11-17 16:09:27,276[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager
2016-11-17 16:09:27,276[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.docker.network=bridge
2016-11-17 16:09:27,276[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dport.http=8080
2016-11-17 16:09:27,277[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.home=/home/user/che
2016-11-17 16:09:27,277[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.logs.dir=/home/user/che/tomcat/logs/
2016-11-17 16:09:27,277[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.logs.level=info
2016-11-17 16:09:27,277[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djdk.tls.ephemeralDHKeySize=2048
2016-11-17 16:09:27,278[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.protocol.handler.pkgs=org.apache.catalina.webresources
2016-11-17 16:09:27,278[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcom.sun.management.jmxremote
2016-11-17 16:09:27,278[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcom.sun.management.jmxremote.ssl=false
2016-11-17 16:09:27,279[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcom.sun.management.jmxremote.authenticate=false
2016-11-17 16:09:27,279[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.local.conf.dir=/data/conf
2016-11-17 16:09:27,279[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.endorsed.dirs=/home/user/che/tomcat/endorsed
2016-11-17 16:09:27,279[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcatalina.base=/home/user/che/tomcat
2016-11-17 16:09:27,280[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcatalina.home=/home/user/che/tomcat
2016-11-17 16:09:27,280[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.io.tmpdir=/home/user/che/tomcat/temp
2016-11-17 16:09:27,368[main]             [INFO ] [o.a.c.http11.Http11NioProtocol 443]  - Initializing ProtocolHandler ["http-nio-8080"]
2016-11-17 16:09:27,387[main]             [INFO ] [o.a.t.util.net.NioSelectorPool 72]   - Using a shared selector for servlet write/read
2016-11-17 16:09:27,391[main]             [INFO ] [o.a.catalina.startup.Catalina 591]   - Initialization processed in 382 ms
2016-11-17 16:09:27,486[main]             [INFO ] [c.m.JmxRemoteLifecycleListener 332]  - The JMX Remote Listener has configured the registry on port 32001 and the server on port 32101 for the Platform server
2016-11-17 16:09:27,486[main]             [INFO ] [o.a.c.core.StandardService 435]      - Starting service Catalina
2016-11-17 16:09:27,487[main]             [INFO ] [o.a.c.core.StandardEngine 259]       - Starting Servlet Engine: Apache Tomcat/8.0.38
2016-11-17 16:09:27,534[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/dashboard.war
2016-11-17 16:09:28,095[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/dashboard.war has finished in 561 ms
2016-11-17 16:09:28,098[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/wsmaster.war
2016-11-17 16:09:38,843[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/wsmaster.war has finished in 10,745 ms
2016-11-17 16:09:38,844[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/swagger.war
2016-11-17 16:09:38,900[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/swagger.war has finished in 56 ms
2016-11-17 16:09:38,902[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/ROOT.war
2016-11-17 16:09:39,787[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/ROOT.war has finished in 885 ms
2016-11-17 16:09:39,794[main]             [INFO ] [o.a.c.http11.Http11NioProtocol 488]  - Starting ProtocolHandler ["http-nio-8080"]
2016-11-17 16:09:39,800[main]             [INFO ] [o.a.catalina.startup.Catalina 642]   - Server startup in 12408 ms

I’m happy to get on a WebEx or equivalent if someone would like to check out my environment. As far as I know, it is vanilla Mac/Docker.

—Darrin

**Comment by l0rd** [comment]
Thanks for all these details @darrinsimmons. I can reproduce your bug now. I'm looking at the code to find the issue.

**Comment by l0rd** [comment]
Ok that's an issue introduced in 5.0.0-M7. The problem is [in this function](https://github.com/eclipse/che/blob/master/assembly/assembly-main/src/assembly/bin/docker.sh#L225): if variable `CHE_DOCKER_MACHINE_HOST_EXTERNAL` is set the function should returns it.

@darrinsimmons as a workaround  you can use version 5.0.0-M6:

``` bash
docker pull eclipse/che-server:5.0.0-M6
docker pull eclipse/che:5.0.0-M6
docker run --rm -t -v /var/run/docker.sock:/var/run/docker.sock \
                               -e CHE_DOCKER_MACHINE_HOST_EXTERNAL=192.168.129.19 \
                               eclipse/che:5.0.0-M6 start
```

I've tested it and it works on my Mac. You probably need to remove your previously created workspaces before testing.

@TylerJewell let me know if you want me to fix it or you can handle this. I would be happy to do that but I won't be able to submit a PR before monday next week.

**Comment by TylerJewell** [comment]
Thanks @l0rd for discovering the issue.  I would very much appreciate you submitting a fix to the docker.sh file, and then I will ask Florent to confirm and test it.  It would make it in for M8.

**Comment by darrinsimmons** [comment]
Thank you @l0rd and @TylerJewell.  When I loaded M6 on my CentOS box, I lost some other functionality (I think), so I'll stick with localhost on my Mac for now until this is fixed.

**Comment by slemeur** [comment]
Closing as there is no more activity on this issue and should have been fixed with #3158
@darrinsimmons : Please reopen if you still have the issue. Thanks
